src/btreemap.rs

```diff
@@ -575,23 +599,35 @@ where
 
     /// Returns the first key-value pair in the map. The key in this
     /// pair is the minimum key in the map.
-    pub fn first_key_value(&self) -> Option<(K, V)> {
+    pub fn first_key_value(&mut self) -> Option<(K, V)> {
         if self.root_addr == NULL {
             return None;
         }
-        let root = self.load_node(self.root_addr);
+
+        if let Some((k, encoded_v)) = &self.first_key_value {
+            return Some((k.clone(), V::from_bytes(Cow::Owned(encoded_v.clone()))));
+        }
+
+        let root: Node<K> = self.load_node(self.root_addr);
         let (k, encoded_v) = root.get_min(self.memory());
+        self.first_key_value = Some((k.clone(), encoded_v.clone()));
         Some((k, V::from_bytes(Cow::Owned(encoded_v))))
     }
 
     /// Returns the last key-value pair in the map. The key in this
     /// pair is the maximum key in the map.
-    pub fn last_key_value(&self) -> Option<(K, V)> {
+    pub fn last_key_value(&mut self) -> Option<(K, V)> {
         if self.root_addr == NULL {
             return None;
         }
+
+        if let Some((k, encoded_v)) = &self.last_key_value {
+            return Some((k.clone(), V::from_bytes(Cow::Owned(encoded_v.clone()))));
```

Comment:
would it be possible to store the actual result of this not doing the work every time?